<html>
<!--
 * Jalview - A Sequence Alignment Editor and Viewer (2.11.3.0)
 * Copyright (C) 2024 The Jalview Authors
 * 
 * This file is part of Jalview.
 * 
 * Jalview is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation, either version 3
 * of the License, or (at your option) any later version.
 *  
 * Jalview is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty 
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
 * PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Jalview.  If not, see <http://www.gnu.org/licenses/>.
 * The Jalview Authors are detailed in the 'AUTHORS' file.
 -->
<title>Command Line: basic usage</title>
<body>

  <h1>Command Line: basic usage</h1>

  <p>
  <a href="clarguments.html">Command Line: introduction</a>
  <br/>
  Command Line: basic usage
  <br/>
  <a href="clarguments-advanced.html">Command Line: advanced usage</a>
  <br/>
  <a href="clarguments-argfiles.html">Command Line: argument files</a>
  <br/>
  <a href="clarguments-reference.html">Command Line: reference</a>
  </p>

  <hr/>

  <ul>
  <li><a href="#openingalignments">Opening alignments</a></li>
  <li><a href="#alignmentoptions">Alignment options</a></li>
  <li><a href="#adding3dstructures">Adding 3D structures</a></li>
  <li><a href="#outputtingalignmentfiles">Outputting/converting alignment files and images</a></li>
  <li><a href="#filenamesubstitutionsandbatchprocessing">Filename substitutions and batch processing</a></li>
  <li><a href="#alloutputwildcard">The all output wildcard</a></li>
  </ul>

  <h2><a name="openingalignments"></a>Opening alignments (<code>--open</code>, <code>--append</code>, <code>--new</code>)</h2>

  <p>
  To simply open one or more alignment files in different windows just put the filenames as the first arguments:
  <pre>
  jalview filename1 filename2 ...
  </pre>
  </p>

  <p>
  You can use shell-expanded wildcards:
  <pre>
  jalview this/filename* that/filename* other/filename*
  </pre>
  and URLs:
  <pre>
  jalview https://rest.uniprot.org/uniprotkb/P00221.fasta
  </pre>
  </p>

  <p>
  (Using initial filenames is the same as using the <code>--open</code> argument, and further arguments can be used
  after the initial filenames.)
  </p>

  <h3><a name="open"></a><code>--open</code></h3>

  <p>
  Use the <code>--open</code> argument to open alignment files each in their own window.
  </p>

  <p>
  The following are equivalent:
  <pre>
  jalview --open filename1 filename2 ...

  jalview --open filename*

  jalview --open filename1 --open filename2 --open ...

  jalview filename1 filename2 ...
  </pre>
  </p>

  <p>
  Similarly you can open URLs:
  <pre>
  jalview --open https://rest.uniprot.org/uniprotkb/P00221.fasta
  </pre>
  </p>

  <h3><a name="append"></a><code>--append</code></h3>

  <p>
  To append several alignment files together use:
  <pre>
  jalview --open filename1.fa --append filename2.fa filename3.fa
  </pre>
  or, if you haven't previously used <code>--open</code> then you can use --append to open one new window and keep appending each set of alignments:
  <pre>
  jalview --append these/filename*.fa --append more/filename*.fa

  jalview --append https://rest.uniprot.org/uniprotkb/P00221.fasta https://www.uniprot.org/uniprotkb/A0A0K9QVB3/entry
  </pre>
  </p>

  <p>
  <strong>Note</strong> that whilst you can include a Jalview Project File (<code>.jvp</code>) as an <code>--append</code> value, the items in the file will always open in their original windows and not append to another.
  </p>

  <h3><a name="new"></a><code>--new</code></h3>

  <p>
  To append different sets of alignment files in different windows, use <code>--new</code> to move on to a new alignment window:
  <pre>
  jalview --append these/filename*.fa --new --append other/filename*.fa
  </pre>
  </p>

  <p>
  <code>--open</code> is like using <code>--new --append</code> applied to every filename/URL given to <code>--open</code>
  </p>


  <h2><a name="alignmentoptions"></a>Alignment options (<code>--colour</code>, <code>--wrap</code>, <code>--showannotations</code>, <code>--title</code>)</h2>

  <h3><a name="colour"></a><code>--colour</code></h3>

  <p>
  You can specify a residue/base colouring for the alignment using the <code>--colour</code> option (note spelling -- Jalview is made in Scotland!):
  <pre>
  jalview --open examples/uniref50.fa --colour gecos-flower
  </pre>
  There are several colour schemes that you can use.  See the <a href="../colourSchemes/index.html">page on Colour Schemes</a> for details.
  The names to use on the command line for colour schemes are:
  </p>
  <p>
  <code>clustal</code>,
  <br/>
  <code>blosum62</code>,
  <br/>
  <code>pc-identity</code>,
  <br/>
  <code>zappo</code>,
  <br/>
  <code>taylor</code>,
  <br/>
  <code>gecos-flower</code>,
  <br/>
  <code>gecos-blossom</code>,
  <br/>
  <code>gecos-sunset</code>,
  <br/>
  <code>gecos-ocean</code>,
  <br/>
  <code>hydrophobic</code>,
  <br/>
  <code>helix-propensity</code>,
  <br/>
  <code>strand-propensity</code>,
  <br/>
  <code>turn-propensity</code>,
  <br/>
  <code>buried-index</code>,
  <br/>
  <code>nucleotide</code>,
  <br/>
  <code>nucleotide-ambiguity</code>,
  <br/>
  <code>purine-pyrimidine</code>,
  <br/>
  <code>rna-helices</code>,
  <br/>
  <code>t-coffee-scores</code>,
  <br/>
  <code>sequence-id</code>
  </p>

  <h3><a name="wrap"></a><code>--wrap</code></h3>
  <p>
  An alignment should open with your usual preferences stored in the <code>.jalview_properties</code> file.  To open an alignment with the sequences (definitely) wrapped, following your <code>--open</code> (or first <code>--append</code>) argument use the argument <code>--wrap</code>:
  <pre>
  jalview --open examples/uniref50.fa --wrap
  </pre>
  To ensure an alignment is not wrapped use <code>--nowrap</code>:
  <pre>
  jalview --open examples/uniref50.fa --nowrap
  </pre>
  </p>

  <h3><a name="showannotations"></a><code>--showannotations</code> / <code>--noshowannotations</code></h3>

  <p>
  You can specify whether the currently opened alignment window should show alignment annotations (e.g. Conservation, Quality, Consensus...) or not with either <code>--showannotations</code> or <code>--noshowannotations</code>.  If you don't specify then your saved preference will be used.
  <pre>
  jalview --open examples/uniref50.fa --noshowannotations
  </pre>
  </p>

  <h3><a name="title"></a><code>--title</code></h3>

  <p>
  If you would like to give the alignment window a specific title you can do so with the <code>--title</code> option:
  <pre>
  jalview --open examples/uniref50.fa --title "My example alignment"
  </pre>
  </p>




  <h2><a name="adding3dstructures"></a>Adding 3D structures (<code>--structure</code>, <code>--seqid</code>, <code>--structureviewer</code>, <code>--paematrix</code>, <code>--tempfac</code>, <code>--showssannotations</code>)</h2>

  <p>
  </p>

  <h3><a name="structure"></a><code>--structure</code></h3>

  <p>
  You can add a 3D structure file to a sequence in the current alignment window with the <code>--structure</code> option:
  <pre>
  jalview --open examples/uniref50.fa --structure examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>
  By default this attaches to the first sequence in the alignment but most likely you will want to attach it to a specific sequence.
  </p>

  <h3><a name="seqid"></a><code>--seqid</code></h3>

  <p>
  The easiest way to specify a sequence ID for your structure is to follow the <code>--structure</code> argument with a <code>--seqid</code> argument with a value of a sequence ID in the alignment.  This does of course require some knowledge of the sequences in the alignment files
  that have been opened.
  <br/>
  Alternatively you can specify a <em>sub-value</em> with the <code>--structure</code> argument value.  You do this by preceding the value with square brackets and <code>seqid=SequenceId</code>,
  like this:
  <pre>
  jalview --open examples/uniref50.fa --structure [seqid=FER1_SPIOL]examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>
  which is equivalent to
  <pre>
  jalview --open examples/uniref50.fa --structure examples/AlphaFold/AF-P00221-F1-model_v4.pdb --seqid FER1_SPIOL
  </pre>
  </p>

  <p>
  The sub-value <code>seqid=FER1_SPIOL</code> takes precedence over the following argument <code>--seqid FER1_SPIOL</code> if you accidentally specify both (in which case the argument will probably be completely unused).
  </p>

  <p>
  If you don't know the sequence IDs but do know the position of the sequence in the alignment, you can also specify an <em>INDEX</em>
  in the sub-values to indicate which sequence in the alignment to attach the sequence to (although this is less precise).  This is a zero-indexed value, so to specify the eighth sequence in the alignment use:
  <pre>
  jalview --open examples/uniref50.fa --structure [7]examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>

  <p>
  Remember that you might need to escape any spaces in the sequence ID or enclose the ID in quotation marks.
  </p>

  <h3><a name="structureviewer"></a><code>--structureviewer</code></h3>

  <p>
  You can specify which structure viewer (or none) to use to open the structure using either the <code>--structureviewer</code> argument or the <code>structureviewer</code> sub-value.  Multiple sub-values can be specified together, separated by a comma ','.  Possible values for the <code>structureviewer</code> are:
  <br/>
  <code>none</code>,
  <br/>
  <code>jmol</code>,
  <br/>
  <code>chimera</code>,
  <br/>
  <code>chimerax</code>,
  <br/>
  <code>pymol</code>.
  </p>
  <p>
  <code>none</code> and <code>jmol</code> will always be available, but to use the others you must have the appropriate software already set up on your computer and in Jalview.  See the page <a href="../features/viewingpdbs.html">Discovering and Viewing PDB and 3D-Beacons structures</a> for more details.
  <pre>
  jalview --open examples/uniref50.fa --structure [seqid=FER1_SPIOL,structureviewer=none]examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>
  or, if you prefer
  <pre>
  jalview --open examples/uniref50.fa --structure examples/AlphaFold/AF-P00221-F1-model_v4.pdb --seqid FER1_SPIOL --structureviewer none
  </pre>
  </p>

  <h3><a name="paematrix"></a><code>--paematrix</code></h3>

  <p>
  If you are opening a structure file that has a PAE matrix (provided as a JSON file), such as from an AlphaFold model or an nf-core pipeline, you can add the PAE matrix as an annotation by following the <code>--structure</code> argument with a <code>--paematrix</code> argument with the filename.  You can also specify a <code>paematrix=filename</code> sub-value.
  <pre>
  jalview --open examples/uniref50.fa --structure [seqid=FER1+SPIOL,structureviewer=pymol]examples/AlphaFold/AF-P00221-F1-model_v4.pdb --paematrix examples/AlphaFold/AF-P00221-F1-predicted_aligned_error_v4.json
  </pre>
  </p>

  <h3><a name="tempfac"></a><code>--tempfac</code></h3>

  <p>
  Structure files may have a temperature factor associated with the structure component positions.  If the temperature factor is a pLDDT confidence score, such as with an AlphaFold model, you can specify this by using a following argument of <code>--tempfac</code> with a value of <code>plddt</code>.  This will enable standard pLDDT colouring of the temperature factor annotation.  Valid values are:
  <code>default</code>,
  <code>plddt</code>.
  More types of temperature factor may be added in future releases of Jalview.
  <br/>
  The value can also be specified as a sub-value:
  <pre>
  jalview --open examples/uniref50.fa --structure [seqid=FER1+SPIOL,structureviewer=jmol,tempfac=plddt]examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>
  which is equivalent to
  <pre>
  jalview --open examples/uniref50.fa --structure examples/AlphaFold/AF-P00221-F1-model_v4.pdb --tempfac plddt --seqid FER1+SPIOL
   --structureviewer jmol
  </pre>

  </p>

  <!-- notempfac not yet working. undocumented until then -->

  <h3><a name="showssannotations"></a><code>--showssannotations</code> / <code>--noshowssannotations</code></h3>

  <p>
  You can specify whether the currently opened alignment window should show secondary structure annotations or not with either <code>--showssannotations</code> or <code>--noshowssannotations</code>.  If you don't specify then your saved preference will be used.
  <pre>
  jalview --open examples/uniref50.fa --structure examples/AlphaFold/AF-P00221-F1-model_v4.pdb --noshowssannotations
  </pre>
  or you can use a sub-value modifier:
  <pre>
  jalview --open examples/uniref50.fa --structure [noshowssannotations]examples/AlphaFold/AF-P00221-F1-model_v4.pdb
  </pre>
  </p>

  <h2><a name="outputtingalignmentfiles"></a>Outputting/converting alignment files and images (<code>--output</code>, <code>--format</code>, <code>--image</code>, <code>--type</code>, <code>--textrenderer</code>, <code>--scale</code>, <code>--backups</code>, <code>--overwrite</code>)</h2>

  <p>
  You can save an alignment as an alignment file, or exported as an image, in different formats.  Jalview's alignment output formats are:
  fasta,
  pfam,
  stockholm,
  pir,
  blc,
  amsa,
  json,
  pileup,
  msf,
  clustal,
  phylip,
  jalview.
  </p>
  <p>
  Alignments can be exported as an image in formats EPS, SVG, HTML, BioJSON (vector formats) or PNG (bitmap format).
  </p>
  <p>
  In vector formats you can specify whether text should be rendered as text (which may have font changes, but will produce a smaller and more usable file) or as lineart (which will retain exact appearance of text, but will be less easy to edit or use to copy text).
  </p>
  <p>
  In bitmap formats (currently only PNG, but what else would you want?!) you can specify a scaling factor to improve the resolution of the output image.
  </p>

  <h3><a name="output"></a><code>--output</code></h3>

  <p>
  To save the open alignment in a new alignment file use <code>--output filename</code>.  The format for the file can be found from the extension of <code>filename</code>, or if you are using a non-standard extension you can use a following <code>--format</code> argument, or specify it as a sub-value modifier.
  </p>
  <p>
  Recognised formats and their recognised extensions are:
  <br/>
  <code>fasta</code> (<code>fa, fasta, mfa, fastq</code>),
  <br/>
  <code>pfam</code> (<code>pfam</code>),
  <br/>
  <code>stockholm</code> (<code>sto, stk</code>),
  <br/>
  <code>pir</code> (<code>pir</code>),
  <br/>
  <code>blc</code> (<code>blc</code>),
  <br/>
  <code>amsa</code> (<code>amsa</code>),
  <br/>
  <code>json</code> (<code>json</code>),
  <br/>
  <code>pileup</code> (<code>pileup</code>),
  <br/>
  <code>msf</code> (<code>msf</code>),
  <br/>
  <code>clustal</code> (<code>aln</code>),
  <br/>
  <code>phylip</code> (<code>phy</code>),
  <br/>
  <code>jalview</code> (<code>jvp, jar</code>).
  <p>
  For example, to open a FASTA file, append another FASTA file and then save the concatenation as a Stockholm file, do
  <pre>
  jalview --open alignment1.fa --append alignment2.fa --output bothalignments.stk
  </pre>
  or
  <pre>
  jalview --append alignment*.fa --output bigballofstring.txt --format stockholm
  </pre>
  or
  <pre>
  jalview --append alignment*.fa --output [format=stockholm]bigballofstring.txt
  </pre>
  </p>

  <p>
  <em>Important!</em> If you use <code>--output</code> or any other argument that outputs a file, then it will be assumed you want to run Jalview in headless mode (as if you had specified <code>--headless</code>).  To use Jalview with <code>--output</code> and not assume headless mode, use the <code>--gui</code> or <code>--noheadless</code> argument (the order doesn't matter).
  </p>

  <h3><a name="format"></a><code>--format</code></h3>

  <p>
  To specify the format of the output file (if using an unrecognised file extension) use the <code>--format</code> argument to specify a value (see above).  A sub-value modifier on the <code>--output</code> value can also be used.
  </p>

  <h3><a name="image"></a><code>--image</code></h3>
  To export the open alignment window as an image, use the <code>--image</code> argument, which will give an image of the alignment and annotations as it appears (or would appear if not in <code>--headless</code> mode) in the alignment window if it was large enough for the whole alignment, including colour choice and features.
  <p>
  <pre>
  jalview --open examples/plantfdx.fa --colour gecos-blossom --features examples/plantfdx.features --annotations examples/plantfdx.annotations --image plantfdx.png --headless
  </pre>
  </p>

  <p>
  This by default produces a PNG image of screen or webpage resolution, which you may want to improve upon.  There are two ways of doing this with Jalview: increasing the scale of the PNG image, or using a vector based image format (EPS, SVG, HTML).
  <p>

  <h3><a name="bitmap"></a>Bitmap image types (<code>png</code>)</h3>

  <p>
  Bitmap images are composed of an array of pixels.  Bitmap images with a low resolution that are blown up to a larger size appear "blocky", so it is important to get the resolution for your purpose correct.  Older screens only require a modest resolution, whilst newer HiDPI screens look better with a higher resolution.  For print you will want an even higher resolution although in this case you would probably want to use a vector image format.  In general creating a bitmap image that has a large resolution means you can scale the image down if needed, although if you are running a batch process this will take more time and resources.
  </p>
  <p>
  Jalview only produces a PNG bitmap image type.  This is a high-colour lossless format which can also use lossless compression so is suitable for alignment figures.
  </p>
  <p>
  Let's increase the resolution of the PNG image:
  </p>

  <h3><a name="scale"></a><code>--scale</code></h3>

  <p>
  We can increase the size of the PNG image by a factor of <em>S</em> by following the <code>--image</code> argument with a <code>--scale <em>S</em></code> argument and value.  The value doesn't have to be an integer and should be given as an integer or floating point formatted number, e.g.
  <pre>
  jalview --open examples/uniref50.fa --colour gecos-ocean --image mypic.png --scale 5.5 --headless
  </pre>
  which will produce a PNG image 5.5 times larger (and more detailed) than without the <code>--scale</code> argument.
  </p>
  <p>
  However, since you won't necessarily already know the size of an alignment's exported image you can specify either an exact width or height (in pixels) with either one of the
  <code>--width</code> and <code>--height</code> arguments:

  <h3><a name="width"></a><code>--width</code></h3>

  <p>
  Specify an exact width of an exported PNG image with <code>--width</code>:
  <pre>
  jalview --headless --open https://www.ebi.ac.uk/interpro/api/entry/pfam/PF03760/?annotation=alignment%3Aseed --noshowannotations --colour gecos-sunset --image wallpaper.png --width 3840
  </pre>
  </p>

  <h3><a name="height"></a><code>--height</code></h3>

  <p>
  Alternatively specify an exact height with the <code>--height</code> argument:
  <pre>
  jalview --headless --open https://www.ebi.ac.uk/interpro/api/entry/pfam/PF03760/?annotation=alignment%3Aseed --noshowannotations --colour gecos-ocean --image wallpaper.png --height 2160
  </pre>
  </p>

  <p>
  You can specify two or all of <code>--scale</code>, <code>--width</code> and <code>--height</code> as limits to the size of the image (think of one or two bounding boxes) and the one which produces the smallest scale of image is used.  You can also specify each of these as sub-value modifiers to the <code>--image</code> value:
  <pre>
  jalview --headless --open https://www.ebi.ac.uk/interpro/api/entry/pfam/PF03760/?annotation=alignment%3Aseed --noshowannotations --colour gecos-flower --image [scale=0.25,width=320,height=240]thumbnail.png
  </pre>
  </p>

  <p>
  Next we look at vector image formats, which maintain detail at all resolutions.
  </p>

  <h3><a name="vector"></a>Vector image export</h3>

  <p>
  Jalview can export an alignment in Encapsulated PostScript (<code>eps</code>), Scalable Vector Graphics (<code>svg</code>), HTML (<code>html</code>) or BioJSON -- another HTML format (<code>biojs</code>), by using, e.g.
  <pre>
  jalview --open examples/uniref50.fa --colour gecos-flower --image printable.eps
  </pre>
  The image format can be specified with the <code>--type</code> argument or as a sub-value modifier on the <code>--image</code> value.  If neither is used the <code>type</code> will be guessed from the image file extension.  The following three examples should produce the same contents:
  <pre>
  jalview --open examples/uniref50.fa --colour gecos-flower --image printable.eps
  jalview --open examples/uniref50.fa --colour gecos-flower --image printable.postscript --type eps
  jalview --open examples/uniref50.fa --colour gecos-flower --image [type=eps]printable.postscript
  </pre>
  </p>

  <h3><a name="textrenderer"></a><code>--textrenderer</code></h3>

  <p>
  In a vector format any text that appears on the page (including residue/base labels) can be saved in the image file either as <code>text</code> or as <code>lineart</code> using the <code>--textrenderer</code> argument.  This is only available for <code>eps</code>, <code>svg</code> and <code>html</code> formats.
  </p>

  <p>
  The difference is potentially in the appearance of the file, and definitely in the filesize!  Saving with <code>text</code> requires the font used to display the text characters in the alignment to be present on the viewing platform to look exactly the same.  If it isn't then another suitable font will probably be used.  The filesize using <code>text</code> is relatively small.
  </p>
  <p>
  When using <code>lineart</code> each individual character that appears in the alignment (including names/titles and resisdues/bases) is stored in the image with its own vector lines.  This means that the appearance of the text is retained exactly independent of installed fonts, but the filesize is increased.  You will also be unable to copy what appears to be text as text.
  </p>

  <p>
  The type of <code>--textrenderer</code> can be specified with an argument following <code>--image</code> or as a sub-value modifier:
  <pre>
  jalview --open examples/uniref50.fa --colour gecos-flower --image printable.html --type biojs
  jalview --open examples/uniref50.fa --colour gecos-flower --image [type=eps,textrenderer=lineart]printable.ps
  </pre>
  </p>


  <h2><a name="filenamesubstitutionsandbatchprocessing"></a>Filename substitutions and batch processing (<code>--substitutions</code>, <code>--close</code>, <code>--all</code>)</h2>

  <p>
  One of the more powerful aspects of Jalview's command line operations is that stores all of the different opened alignment arguments (before opening them) and can apply some arguments to <em>all</em> of the alignments as they are opened.  This includes display and output arguments.
  </p>

  <p>
  When outputting a file you will obviously want to use a different filename for each of the alignments that are opened.  There are several ways you can specify how that filename differs, but the easiest way is to refer to the input filename and change the extension.  In the case of an alignment where multiple files are appended, the filename of the first file to be appended or opened is used.
  </p>

  <p>
  To refer to different parts of the opening filename, you can use the strings
  <ul>
  <li><code>{dirname}</code> -- is replaced by the directory path to the opened file.</li>
  <li><code>{basename}</code> -- is replaced by the base of the filename of the opened file. This is without the path or file extension (if there is one).</li>
  <li><code>{extension}</code> -- is replaced by the extension of the filename of the opened file.</li>
  </ul>
  The braces (curly brackets '{', '}') are important!
  </p>

  <p>
  These filename substitutions are on by default, but if for whatever reason you wish to disable the substitutions, they can be turned off (or back on again) through the list of arguments with:
  </p>

  <h3><a name="substitutions"></a><code>--substitutions / --nosubstitutions</code></h3>

  <p>
  Enable (or disable) filename substitutions in the following argument values and sub-value modifier values.
  <pre>
  jalview --open exampes/uniref50.fa --nosubstitutions --output I_Want_A_Weird_Output_Filname_With_{basename}_Curly_Brackets_In_And_A_Sensible_One_Next.stk --substitutions --output tmp/{basename}.stk --headless
  </pre>
  </p>

  <p>
  For opening single files this is less useful, since you could obviously just type the output filename, but for multiple opened alignments you can also use these substituted values and they will be replaced by the relevant part of the filename given for each opened alignment window.  Normally an <code>--output</code> or <code>--image</code> argument will only be applied to the latest opened alignment window, but you can tell Jalview to apply some arguments to all alignments that have been opened (so far) by using the <code>--all</code> argument.
  </p>

  <h3><a name="all"></a><code>--all / -noall</code></h3>

  <p>
  When using the <code>--all</code> argument, following arguments will apply to all of the previously opened alignment windows.  You can turn this behaviour off again for following arguments using the <code>--noall</code> argument.  The arguments that can apply to all previously opened alignments are:
  <br/>
  <code>--colour</code>
  <br/>
  <code>--sortbytree</code>
  <br/>
  <code>--showannotations</code>
  <br/>
  <code>--wrap</code>
  <br/>
  <code>--nostructure</code>
  <br/>
  <code>--notempfac</code>
  <br/>
  <code>--showssannotations</code>
  <br/>
  <code>--image</code>
  <br/>
  <code>--type</code>
  <br/>
  <code>--textrenderer</code>
  <br/>
  <code>--scale</code>
  <br/>
  <code>--width</code>
  <br/>
  <code>--height</code>
  <br/>
  <code>--output</code>
  <br/>
  <code>--format</code>
  <br/>
  <code>--groovy</code>
  <br/>
  <code>--backups</code>
  <br/>
  <code>--overwrite</code>
  <br/>
  <code>--close</code>
  </p>
  <p>
  In particular, for our example above, we can use <code>-all</code> and <code>--output</code> like this (<code>--close</code> will be explained in a moment):
  <pre>
  jalview --open all_my_fasta_files/*.fa --all --output all_my_converted_stockholm_files/{basename}.stk --close --headless
  </pre>
  </p>

  <p>
  Another example would be to create a print quality coloured postscript image for every Fasta file in several directories and place the image next to the alignment:
  <pre>
  jalview --open */*.fa --all --colour gecos-flower --image {dirname}/{basename}.eps --textrenderer text --close --headless
  </pre>
  or thumbnails for every Fasta file with
  <pre>
  jalview --open */*.fa --all --colour gecos-flower --image {dirname}/{basename}.png --width 256 --height 256 --close --headless
  </pre>
  </p>

  <h3><a name="close"></a><code>--close</code></h3>

  <p>
  The <code>--close</code> tag is used to close an alignment window after all output files or image exports are performed.  This reduces memory use, especially if an <code>--open</code> value is set to open many files.  These will be opened, formatted and output sequentially so since they are closed before the next one is opened memory use will not build up over a large number of alignments.
  <pre>
  </pre>
  </p>


  <h2><a name="alloutputwildcard"></a>The all output wildcard: <code>--output "*.ext"</code>,  <code>--image "*.ext"</code></h2>

  <p>
  Purely as an intuitive syntactic sweetener, you can use the <code>--output</code> wildcard <code>*</code> <em>at the beginning of the output filename</em> as shorthand for <code>--all --output {dirname}/{basename}</code> followed by whatever you put after the '<code>*</code>'.  For example, to achieve the same as the thumbnails example above, you could use
  <pre>
  jalview --open */*.fa --image "*.png" --colour gecos-flower --width 256 --height 256 --close --headless
  </pre>
  Here we move the <code>--colour</code> argument after the <code>--output</code> argument (it will still be applied before the image export or file output) so that it is included after the implied <code>--all</code> argument.  The thumbnails will be placed in the same directory as the alignment file with the same filename except for a different extension of <code>.png</code>.
  </p>

  <p>
  For a simple conversion of many fasta files into Stockholm format, simply use
  <pre>
  jalview --open all_my_fasta_files/*.fa --output "*.stk" --close --headless
  </pre>
  </p>

  <p>
  <strong>Important!</strong>  Please note that using a bareword <code>*.ext</code> (i.e. without an escape or quotation marks) in a shell command line will potentially be expanded by the shell to a list of all the files in the current directory ending with <code>.ext</code> <em>before</em> this value is passed to Jalview.  This will result in the first of these files being overwritten (multiple times)!  If you shell-escape the <code>*</code> (usually with a backslash '\') or enclose the value in quotation marks (<code>"*.ext"</code> - that's including the quotation marks) then the shell will not expand the wildcard.
  </p>

  <p>
  An alternative is to use an equals sign ('=') with no spaces between the argument and the value, <code>--output=*.ext</code>, which Jalview will interpret the same, but the shell will not automatically expand before it is sent to Jalview, e.g.
  <pre>
  jalview --open all_my_fasta_files/*.fa --output=*.stk --close --headless
  </pre>
  </p>

  <hr/>
  Continue to <a href="clarguments-advanced.html">Command Line: advanced usage</a>.
  <br/>
  <a href="clarguments-reference.html">Command Line: reference</a>

</body>
</html>
